Meta/CMake/Superbuild/CMakeLists.txt

```diff
@@ -0,0 +1,79 @@
+cmake_minimum_required(VERSION 3.16)
+
+# Variables that we really don't want the user to pass to the serenity cross-build because they will screw things up :^)
+set(LAGOM_ONLY_VARS "CMAKE_CXX_COMPILER;CMAKE_C_COMPILER")
+set(EXCLUDE_FROM_CL_VARS "SERENITY_ARCH;SERENITY_TOOLCHAIN")
+
+# https://stackoverflow.com/a/48555098
+# https://cmake.org/pipermail/cmake/2018-January/067002.html
+# MUST be done before call to 'project'
+get_cmake_property(vars CACHE_VARIABLES)
```

Comment:
This is what I meant with my comment in the CI workflow. Provide the superbuild with a narrow contract and just document its usage. Are really that many variables shared by both that this approach is needed?

Reply:
There are about 10 options that are supported by both. Essentially:

- Sanitizer enable flags
- compile time format string flag
- enable all debug flag/all debug facilities flag for CI to ensure we don't have stale code behind -DFOO_DEBUG=1 (which was a problem I'm the past, and the reason all_the_debug_macros.cmake exists in the first place)
- flags that control downloading and using external resources. Ex. Wasm spec tests, Unicode character database, Unicode calendar database, etc. Still need to update the branch fully for the addition of the calendar database last week.

My biggest concerns are:

1) Some devs, esp those on macOS but some on older LTS Linux distros as well, have to pass a CMAKE_CXX_COMPILER today because their default host c++ compiler is not new enough to compile Lagom. Mostly AK/Format.h to be honest, though there's some concepts sprinkled around everywhere :). I really don't want them setting it for the superbuild configure though, or it'll get passed to the target build and screw everything up. Maybe that's a problem I created for myself by moving the Toolchain generation to the serenity build? The less options someone has the specify on the command line, the better imo :). 

2) It's unclear what the best way to tell CMake ok use your previous settings but add these and do the thing again. For example, if I just want to enable UBSAN on target, but keep everything else the same, how do I do that only interacting with the target build directory? I don't want to have to re-specify the CMake install prefix or common downloaded Unicode data file directory just to add ENABLE_UNDEFINED_SANITIZER=ON, or to set -DENABLE_EXTRA_KERNEL_DEBUG to get -O0 -ggdb3 for the kernel.

Reply:
Oh, and 3.

I really don't want to have to choose between visibility of options, and maintenance of multiple copies of the same option. I guess that's what Documentation/AdvancedBuildSettings.md (name?) Is for... But normally I remind myself what the options are by checking CMakeLists.txt. would be a shame to have to do it by checking common_options.cmake, then serenity_options.cmake or some such. Maybe that's a better design in general though 🤷‍♂️

Reply:
> Some devs [...] have to pass a CMAKE_CXX_COMPILER today because their default host c++ compiler is not new enough to compile Lagom

The more convenient way to achieve this is to use the [CC](https://cmake.org/cmake/help/latest/envvar/CC.html) and [CXX](https://cmake.org/cmake/help/latest/envvar/CXX.html) env variables, which applies to more tools than just CMake as well.  
I regularly pass these variables to different builds without a second thought, so I guess this is more of a mindset kind of thing. I'm pondering whether more documentation would help here, but I'm still on the side of a narrow contract here.

> The less options someone has the specify on the command line, the better imo

Debatable. It can quickly become a slippery slope and you end up with a lot of hardcoded details that are now impossible to override from the CLI. Care must be taken to not end up there.

> It's unclear what the best way to tell CMake ok use your previous settings but add these and do the thing again

That's what the CMake cache is for:

```sh
# Let's build shared libs with ccache
cmake -S . -B build -G Ninja -D BUILD_SHARED_LIBS=1 -D CMAKE_CXX_COMPILER_LAUNCHER=ccache
# Actually, I wanted static libs, still build with ccache
cmake build -D BUILD_SHARED_LIBS=0
# Nope, need shared, with ccache, also export compile commands
cmake build -D BUILD_SHARED_LIBS=1 -D CMAKE_EXPORT_COMPILE_COMMANDS=1
```

> I really don't want to have to choose between visibility of options, and maintenance of multiple copies of the same option.

You won't have visibility of options of child builds from the superbuild. You can `ccmake`/`cmake-gui` the individual child builds to insecpt their cache for the variables. Maybe I'm misundersting you here.

Reply:
I didn't know that you could modify the CMake cache like that with `cmake <build_dir> <cache var to change>`. So, armed with that knowledge, it makes a heck of a lot more sense to use the Toolchain file properly instead of trying my hacky approach. I'll revert those hacks and move the cmake toolchain file generation into the superbuild instead, like I did the first go-around.

Reply:
Regarding the common options, you can extract them to a cmake script:

```cmake
if(NOT COMMAND serenity_variable)
  macro(serenity_variable)
    set(${ARGV})
  endmacro()
endif()

serenity_variable(FOO BAR CACHE STRING "Documentaiton, ho!")
# ...
```

and in the superbuild, you can define the `serenity_variable` command ahead of including this, so you can collect the variables and their value automatically:

```cmake
set(common_options "")
macro(serenity_variable name value)
  set(${ARGV})
  list(APPEND common_options "-D${name}:STRING=${value}")
endmacro()

include(.../common-options.cmake)
```

This way you can list expand the `common_options` variable to either `CMAKE_CACHE_ARGS` or `CMAKE_CACHE_DEFAULT_ARGS`.

Reply:
that sounds good, but in order to let the cache for the superbuild affect the variable, I had to do this, with "-D${name}:STRING:=${${name}}"
```
# Collect options for Lagom build
set(lagom_options "")
macro(serenity_option name value)
    set(${ARGV})
    list(APPEND lagom_options "-D${name}:STRING=${${name}}")
endmacro()
```
push incoming, without a rebase :)

Reply:
Yes, you are correct. I was in a bit of a rush to put that together. No need for the `value` parameter in that case.